### PR TITLE
Prepend the namespace to ironic nodenames to avoid conflicts

### PR DESCRIPTION
**Pull request by asalkeld**
Modify the hostname for ironic to prepend the namespace to the name to avoid naming conflicts.

### CODE REVIEW

File: pkg/provisioner/ironic/ironic.go [hunk 1]
```diff
@@ -198,9 +198,9 @@ func newProvisionerWithIronicClients(host metal3v1alpha1.BareMetalHost, bmcCreds
 		bmcCreds:  bmcCreds,
 		client:    clientIronic,
 		inspector: clientInspector,
-		log:       log.WithValues("host", host.Name),
 		publisher: publisher,
 	}
+	p.log = log.WithValues("host", p.hostName())
```
Comment (hardys) [inline pkg/provisioner/ironic/ironic.go:203]: Was there a reason this needed to move from the ironicProvisioner arguments?

File: pkg/provisioner/ironic/ironic.go [hunk 2]
```diff
@@ -1784,6 +1784,13 @@ func (p *ironicProvisioner) softPowerOff() (result provisioner.Result, err error
 	return result, nil
 }
 
+func (p *ironicProvisioner) hostName() string {
+	if p.host.Namespace == "openshift-machine-api" {
```
Comment (hardys) [inline pkg/provisioner/ironic/ironic.go:1788]: IMO we shouldn't hardcode this, it's specific to the openshift deployment of the BMO.

Could we perhaps only prepend the namespace in the case where [WATCH_NAMESPACE](https://github.com/metal3-io/baremetal-operator/blob/master/main.go#L89) is not set?

In the Openshift case, can we restart the ironic containers when the single->multi-namespace [reconfiguration via the CBO](https://github.com/openshift/cluster-baremetal-operator/pull/106) happens?? /cc @sadasu @dhellmann 

If not then I guess another option is to look for an existing host (e.g the `findExistingHost()` function - it's not clear to me if we ever do the lookup by name in practice, but that function could be made to look for both the old/new name format?

Comment (dhellmann) [reply pkg/provisioner/ironic/ironic.go:1788]: > IMO we shouldn't hardcode this, it's specific to the openshift deployment of the BMO.

I agree, we don't want OpenShift-specific behavior hardcoded upstream.

> Could we perhaps only prepend the namespace in the case where [WATCH_NAMESPACE](https://github.com/metal3-io/baremetal-operator/blob/master/main.go#L89) is not set?

We've assumed that the name field in ironic needs to be unique. Is that actually true? How does upstream deal with the potential for name conflicts when watching all namespaces?

If the name does need to be unique, I think I would rather always have the name prefix (rather than triggering that behavior based on whether we're only watching a namespace, unless that proves difficult to implement.

> In the Openshift case, can we restart the ironic containers when the single->multi-namespace [reconfiguration via the CBO](https://github.com/openshift/cluster-baremetal-operator/pull/106) happens?? /cc @sadasu @dhellmann

In OpenShift, reconfiguring metal3 will restart the whole pod, yes.

> If not then I guess another option is to look for an existing host (e.g the `findExistingHost()` function - it's not clear to me if we ever do the lookup by name in practice, but that function could be made to look for both the old/new name format?

I honestly don't remember the use case for looking up the node in ironic by its name instead of UUID. Do we need that behavior, or was it one of my belt-and-braces choices?

On upgrade, an existing host should find it's ironic node via the UUID. Can we change the name of an existing node in ironic to match the new pattern?

Comment (asalkeld) [reply pkg/provisioner/ironic/ironic.go]: So it should be better now (not hardcoded)..

Comment (hardys) [reply pkg/provisioner/ironic/ironic.go]: > We've assumed that the name field in ironic needs to be unique. Is that actually true? How does upstream deal with the potential for name conflicts when watching all namespaces?

/cc @dtantsur but I think it does need to be unique, because the API supports lookup via both uuid and name node_ident (and because the BMO code does a lookup by name when it fails to find the UUID...)

I think upstream the `owner` field must be used for each tenant/project ID - so perhaps we should do the same?

https://docs.openstack.org/api-ref/baremetal/?expanded=create-node-detail

> On upgrade, an existing host should find it's ironic node via the UUID. Can we change the name of an existing node in ironic to match the new pattern?

Yes this does appear to work, if we decide not to use the owner field:

```
$ baremetal node list | grep worker-0
| 7c5d5dcb-cd32-4538-90b3-aff8523e16f4 | ostest-worker-0 | 547a2958-b1f6-4904-8c56-5a6516936e62 | None        | active             | True        |

$ baremetal node set --name testns-ostest-worker-0 ostest-worker-0
Management cluster forced to be minikube when container runtime is not docker

$ baremetal node list | grep worker-0
| 7c5d5dcb-cd32-4538-90b3-aff8523e16f4 | testns-ostest-worker-0 | 547a2958-b1f6-4904-8c56-5a6516936e62 | None        | active             | True        |
```

Comment (dtantsur) [reply pkg/provisioner/ironic/ironic.go]: Nodes names are unique in ironic, as you've already checked. `owner` is not a namespace, names have to be globally unique. Otherwise you can use `owner` as you wish, it only has special meaning with keystone.

**Review by dhellmann — COMMENTED** [review]
It's a little unfortunate that we have to leak the deployment configuration into the provisioner, but I don't see an elegant way to avoid it that doesn't leak the implementation of the provisioner up into `main()`.

File: pkg/provisioner/ironic/ironic.go [hunk 3]
```diff
@@ -1784,6 +1790,14 @@ func (p *ironicProvisioner) softPowerOff() (result provisioner.Result, err error
 	return result, nil
 }
 
+func (p *ironicProvisioner) hostName() string {
+	if !p.watchAllNamespaces || p.host.Namespace == p.podNamespace {
+		return p.host.Name
+	}
+
+	return p.host.Namespace + "-" + p.host.Name
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply pkg/provisioner/ironic/ironic.go:1798]: ok, so would the "." be better? I thought it's a "hostname" and a "." might be seen as a sub-domain - but maybe this is not an issue.

[the remark this answers is not in the document]
Comment (dhellmann) [reply pkg/provisioner/ironic/ironic.go:1798]: What about `/`? Maybe ironic doesn't allow that?

File: config/manager/manager.yaml [hunk 4]
```diff
@@ -27,6 +27,10 @@ spec:
             valueFrom:
               fieldRef:
                 fieldPath: metadata.name
+          - name: POD_NAMESPACE
+            valueFrom:
+              fieldRef:
+                fieldPath: metadata.namespace
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply config/manager/manager.yaml:33]: @dhellmann really? AFAIK it is an automatic variable that picks up the namespace from the pod that is running it.
https://kubernetes.io/docs/tasks/inject-data-application/environment-variable-expose-pod-information/

[the remark this answers is not in the document]
Comment (asalkeld) [reply config/manager/manager.yaml:33]: doh, forgot it is manually defined: https://github.com/openshift/cluster-baremetal-operator/blob/dd41e00d29316cfd4a86334f63825e3279fb1715/provisioning/baremetal_pod.go#L283-L302

Here is a PR for that https://github.com/openshift/cluster-baremetal-operator/pull/107

File: pkg/provisioner/ironic/ironic.go (hunk 2 above)
Comment (hardys) [reply pkg/provisioner/ironic/ironic.go]: > names have to be globally unique

Thanks @dtantsur - how does OpenStack handle this then, in the case where you have Nova instances per-project which don't have globally unique names?

If the name has to be globally unique then I don't think the `owner` field helps us, so the current prepending approach is probably our only option?

Comment (dtantsur) [reply pkg/provisioner/ironic/ironic.go]: Nova instance names are not related to node names, Nova matches ironic nodes by UUID (more precisely, hypervisor UUID which is equal to node UUID for Ironic).

File: pkg/provisioner/ironic/ironic.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (dtantsur) [reply pkg/provisioner/ironic/ironic.go]: I think we expect it to be a valid hostname, I don't think we allow `/`

File: pkg/provisioner/ironic/ironic.go (hunk 2 above)
Comment (dtantsur) [reply pkg/provisioner/ironic/ironic.go]: In the ironic world, an instance name is rather mapped to an allocation name (I don't think we use allocation in BMO, but we might). The node name is something descriptive of the node itself (rack-7-node-32).

Comment (dhellmann) [reply pkg/provisioner/ironic/ironic.go]: I don't know what an allocation name is, so we probably don't use it.

The BMO should always be able to look up a node by UUID, although we did consider other options for discovery. The name is used as a fallback, and as a debugging aid.

What if we combine the host CR UUID, namespace, and name? That would give us a unique value that couldn't be spoofed and we can use whatever punctuation separators are valid without concerns about collisions.

File: config/manager/manager.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (sadasu) [reply config/manager/manager.yaml:33]: Should Namespace passed in via POD_NAMESPACE be provided for BMO Leader election here: https://github.com/metal3-io/baremetal-operator/blob/master/main.go#L114?

File: pkg/provisioner/ironic/ironic.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (asalkeld) [reply pkg/provisioner/ironic/ironic.go]: @dtantsur what character would you suggest?

File: pkg/provisioner/ironic/ironic.go [hunk 5]
```diff
@@ -1784,6 +1790,14 @@ func (p *ironicProvisioner) softPowerOff() (result provisioner.Result, err error
 	return result, nil
 }
 
+func (p *ironicProvisioner) hostName() string {
+	if !p.watchAllNamespaces || p.host.Namespace == p.podNamespace {
+		return p.host.Name
+	}
+
+	return p.host.Namespace + "." + p.host.Name
```
Comment (maelk) [inline pkg/provisioner/ironic/ironic.go:1798]: Would this break the behaviour during upgrade ? If the nodes are created with an old version of BMO, they only have their name, but then we might be looking for it with namespace + name after the upgrade.

File: pkg/provisioner/ironic/ironic.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (hardys) [reply pkg/provisioner/ironic/ironic.go]: Note that we pass the BMH (not Ironic) name as metadata to the host:

https://github.com/metal3-io/baremetal-operator/blob/master/pkg/provisioner/ironic/ironic.go#L1399

So although we'll need to align with the ironic validation constraints, I don't think this name actually ever gets used as a hostname on the target OS?

File: pkg/provisioner/ironic/ironic.go (hunk 5 above)
Comment (hardys) [reply pkg/provisioner/ironic/ironic.go:1798]: I think it depends if we expect to support upgrading from an environment without watchAllNamespaces to one where it's true - there was some discussion about that in the OpenShift case here:

https://github.com/openshift/cluster-baremetal-operator/pull/106#discussion_r576242689

@sadasu @dhellmann did we reach a conclusion on that discussion?

If we expect to support upgrading from a single-namespace to an all-namespaces configuration, I think we'll need to modify the conditional in ValidateManagementAccess where we rename the Ironic host only when `ironicNode.Name == ""` - we could rename whenever the current hostname doesn't match `p.hostName()` there instead?

File: config/manager/manager.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (asalkeld) [reply config/manager/manager.yaml:33]: @sadasu I don't think that has to change.

File: pkg/provisioner/ironic/ironic.go (hunk 5 above)
Comment (asalkeld) [reply pkg/provisioner/ironic/ironic.go:1798]: ok, i'll make that change.

File: pkg/provisioner/ironic/ironic.go [hunk 6]
```diff
@@ -1784,6 +1790,14 @@ func (p *ironicProvisioner) softPowerOff() (result provisioner.Result, err error
 	return result, nil
 }
 
+func (p *ironicProvisioner) hostName() string {
```
Comment (andfasano) [inline pkg/provisioner/ironic/ironic.go:1793]: I'd suggest to rename this method to something like ie `getIronicNodeNameFromHost(host metal3v1alpha1.BareMetalHost)` to make it more clear that this is mapping function to calculate the effective Ironic node name to be used given a BMH (since the current name sounds a little bit too vague for me)

Comment (asalkeld) [reply pkg/provisioner/ironic/ironic.go:1793]: I'll meet you half way, I still need p.watchAllNamespaces, so I'll this
```
func (p *ironicProvisioner) IronicNodeNameFromHost() string
```
also golang getters shouldn't start with "get" > https://golang.org/doc/effective_go#Getters

File: pkg/provisioner/ironic/ironic.go [hunk 7]
```diff
@@ -192,15 +196,17 @@ func newProvisionerWithIronicClients(host metal3v1alpha1.BareMetalHost, bmcCreds
...
-	}
+		host:               host,
+		status:             &(host.Status.Provisioning),
+		bmcAccess:          bmcAccess,
+		bmcCreds:           bmcCreds,
+		client:             clientIronic,
+		inspector:          clientInspector,
+		publisher:          publisher,
+		watchAllNamespaces: os.Getenv("WATCH_NAMESPACE") == "",
+		podNamespace:       os.Getenv("POD_NAMESPACE"),
```
Comment (andfasano) [inline pkg/provisioner/ironic/ironic.go:207]: Is it possible to pass these two params along with the other params in the factory method? Otherwise this strictly coupling could affect also the unit testing.

File: pkg/provisioner/ironic/ironic.go (hunk 6 above)
Comment (andfasano) [reply pkg/provisioner/ironic/ironic.go]: Mine was just an example, `IronicNodeNameFromHost` it's fine for me :+1:

**Review by sadasu — COMMENTED** [review]
/lgtm

File: pkg/provisioner/ironic/ironic.go [hunk 8]
```diff
@@ -1784,6 +1790,14 @@ func (p *ironicProvisioner) softPowerOff() (result provisioner.Result, err error
 	return result, nil
 }
 
+func (p *ironicProvisioner) ironicNodeNameFromHost() string {
+	if !p.watchAllNamespaces || p.host.Namespace == p.podNamespace {
```
Comment (sadasu) [inline pkg/provisioner/ironic/ironic.go:1794]: If we are watching all namespaces and a BMH resource is created in pod namespace, will this return just the hostname (without the namespace prefix)? 
Even in that scenario, if I understand this feature correctly, we would want this method to return <namespace>.<hostname> whenever p.watchAllNamespaces.

**Review by sadasu — COMMENTED** [review]
Do we have to consider the disable case, where BMO goes from watching all namespaces back to just watching 1 namespace? In that case would it be better for the method that generates the node name to always generate the name with the Namespace prefix?

File: pkg/provisioner/ironic/ironic.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (imain) [reply pkg/provisioner/ironic/ironic.go]: I wonder about using a UUID for the hostname?  Sounds like we don't use it for lookup so it shouldn't affect that.  This would mess with debuggability a bit but could make this all a lot easier.

**Comment by imain** [comment]
lgtm!

**Review by imain — APPROVED** [review]
Looks good to me

File: pkg/provisioner/ironic/ironic.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (hardys) [reply pkg/provisioner/ironic/ironic.go]: > I wonder about using a GUID for the hostname? Sounds like we don't use it for lookup so it shouldn't effect that. This would mess with debuggability a bit but could make this all a lot easier.

Personally I think it would affect debuggability a lot, as a developer there's a lot of value in being able to interact with the Ironic API directly and be able to see at a glance which hosts align with specific BMH resources.

File: pkg/provisioner/ironic/ironic.go (hunk 8 above)
Comment (asalkeld) [reply pkg/provisioner/ironic/ironic.go:1794]: so the point, as I understand it, is to prevent these names from clashing (because they are different clusters and you could have nodes with the same name). The idea is that if we go from watchAllNamespace to !watchAllNamespace and back at least the current cluster (it's bmh is in POD_NAMESPACE) would always be the same. I think @hardys suggested keeping the current namespaced names without the "." so that when we upgrade the names are consistent.

Comment (sadasu) [reply pkg/provisioner/ironic/ironic.go:1794]: When BMH resources are listed for all namespaces, the output would contain some with a namespace prefix and others without. Just making sure we are ok with it.

Comment (hardys) [reply pkg/provisioner/ironic/ironic.go:1794]: In the thread below I wasn't suggesting we keep the current names, but that we should rename existing ironic nodes in the case where the configuration has gone from single-namespace to watchAllNamespaces.

E.g here https://github.com/metal3-io/baremetal-operator/blob/master/pkg/provisioner/ironic/ironic.go#L493..L498 we'd change the name whenever it doesn't match the expected format.

It'll need some testing to ensure there aren't corner cases to consider, but that would resolve this concern I think?

Obviously the code here would also need to return the prefixed name whenever the watchAllNamespaces mode is enabled.

Comment (hardys) [reply pkg/provisioner/ironic/ironic.go:1794]: > The idea is that if we go from watchAllNamespace to !watchAllNamespace and back 

I was thinking that the only sane path was to switch from a single-namespace to `watchAllNamespaces=true` - that was the focus of my previous comment, I don't see how we can really support the reverse path, since there could be BMH resources in any namespace?

Comment (asalkeld) [reply pkg/provisioner/ironic/ironic.go:1794]: Thanks for the clarifcation @hardys , I'll fixup today.

File: pkg/provisioner/ironic/ironic.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (imain) [reply pkg/provisioner/ironic/ironic.go]: Yeah good point.

File: pkg/provisioner/ironic/ironic.go (hunk 2 above)
Comment (zaneb) [reply pkg/provisioner/ironic/ironic.go]: The reason we look up by name is as a fallback if we fail to record the UUID of the node in the Status after creating it; it's necessary to have such a fallback to make the creation idempotent.
We do use the allocation ID, it's set to the UID of the Host. I don't think it's set immediately on creation though.

File: pkg/provisioner/ironic/ironic.go [hunk 9]
```diff
@@ -192,22 +194,23 @@ func newProvisionerWithIronicClients(host metal3v1alpha1.BareMetalHost, bmcCreds
...
-		log:       log.WithValues("host", host.Name),
-		publisher: publisher,
+		host:               host,
+		status:             &(host.Status.Provisioning),
+		bmcAccess:          bmcAccess,
+		bmcCreds:           bmcCreds,
+		client:             clientIronic,
+		inspector:          clientInspector,
+		publisher:          publisher,
+		watchAllNamespaces: watchAllNamespaces,
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply pkg/provisioner/ironic/ironic.go:204]: sure, changing..

**Comment by furkatgofurov7** [comment]
There was [an issue](https://kubernetes.slack.com/archives/CHD49TLE7/p1615895171009200) with our CI recently, so tests did not run properly in the previous runs. Hopefully should be fine now.

/test-integration

File: pkg/provisioner/ironic/ironic.go [hunk 10]
```diff
@@ -303,18 +304,20 @@ func (p *ironicProvisioner) findExistingHost() (ironicNode *nodes.Node, err erro
...
-	case nil:
-		p.log.Info("found existing node by name")
-		return ironicNode, nil
-	case gophercloud.ErrDefault404:
-		p.log.Info(
-			fmt.Sprintf("node with name %s doesn't exist", p.host.Name))
-	default:
-		return nil, errors.Wrap(err,
-			fmt.Sprintf("failed to find node by name %s", p.host.Name))
+	for _, nodeName := range []string{p.ironicNodeNameFromHost(), p.host.Name} {
```
Comment (zaneb) [inline pkg/provisioner/ironic/ironic.go:307]: Thinking a bit more about this, I think I was wrong and we don't reeeeally need the fallback in practice.

* If the Ironic DB is in ephemeral storage (almost everyone) then when you update all of the old nodes will be lost anyway. The only exception would be if you change the config but the Ironic DB is in a different Pod that doesn't get restarted.
* If it's in persistent storage, you won't run into this except for Hosts that were just created in the middle of switching from namespaced->clusterwide and which failed to write their ProvisionerID to the k8s API.

However, there is a real security issue here. If you have RBAC permissions to create a BareMetalHost, even in a namespace where it is not reconciled, then you could create a Host with the name `admin-namespace~admin-host` in advance of the switch from single-namespace to cluster-wide. Then you will end up with a BareMetalHost that has the same ProvisionerID as the one in the admin namespace, and thus with some luck and maybe timing you will be able to control it without having the credentials.

We could eliminate the security issue by not falling back in cases where the host name contains a `~`.

File: main.go [hunk 11]
```diff
@@ -111,7 +111,7 @@ func main() {
 		Port:                    0, // Add flag with default of 9443 when adding webhooks
 		LeaderElection:          enableLeaderElection,
 		LeaderElectionID:        "baremetal-operator",
-		LeaderElectionNamespace: watchNamespace,
+		LeaderElectionNamespace: os.Getenv("POD_NAMESPACE"),
```
[the remark this answers is not in the document]
Comment (asalkeld) [reply main.go]: sure, did that

File: pkg/provisioner/ironic/ironic.go (hunk 10 above)
Comment (asalkeld) [reply pkg/provisioner/ironic/ironic.go]: done

File: pkg/provisioner/ironic/ironic.go [hunk 12]
```diff
@@ -303,18 +304,25 @@ func (p *ironicProvisioner) findExistingHost() (ironicNode *nodes.Node, err erro
...
-		p.log.Info("found existing node by name")
-		return ironicNode, nil
-	case gophercloud.ErrDefault404:
-		p.log.Info(
-			fmt.Sprintf("node with name %s doesn't exist", p.host.Name))
-	default:
-		return nil, errors.Wrap(err,
-			fmt.Sprintf("failed to find node by name %s", p.host.Name))
+	nodeSearchList := []string{p.ironicNodeNameFromHost()}
+	if !strings.Contains(p.host.Name, "~") {
```
Comment (zaneb) [inline pkg/provisioner/ironic/ironic.go:308]: nit: we should really use the constant nameSeparator here.

**Review by zaneb — COMMENTED** [review]
One tiny comment.
/approve

File: pkg/provisioner/ironic/ironic.go (hunk 12 above)
Comment (asalkeld) [reply pkg/provisioner/ironic/ironic.go]: doh, fixed

**Comment by zaneb** [comment]
This has been around for a while and has had quite a few eyes on it.
/lgtm

**Comment by asalkeld** [comment]
needed a unit test fix after @honza's PR got merged.
